File: program/item_key.py
"""Стабильный ключ пункта подборки для хранения оценок."""

from __future__ import annotations

import hashlib
import re

_MAX_TEXT = 500


def normalize_item_text(text: str) -> str:
    return re.sub(r"\s+", " ", text.strip().lower())


def make_item_key(section: str, text: str) -> str:
    """Хеш содержимого пункта — не зависит от индекса и версии программы."""
    payload = f"{section}:{normalize_item_text(text)}"
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16]


def make_route_stop_key(poi_id: str) -> str:
    """Стабильный ключ оценки POI-остановки (обратимый, не хеш)."""
    pid = (poi_id or "").strip()
    if not pid:
        raise ValueError("poi_id обязателен")
    return f"poi:{pid}"


def parse_route_stop_key(item_key: str) -> str | None:
    key = (item_key or "").strip()
    if key.startswith("poi:") and len(key) > 4:
        return key[4:]
    return None

File: api-node/src/lib/itemKey.ts
import { createHash } from "node:crypto";

export function makeItemKey(section: string, text: string): string {
  const normalized = text.trim().toLowerCase().replace(/\s+/g, " ");
  const payload = `${section}:${normalized}`;
  return createHash("sha256").update(payload, "utf8").digest("hex").slice(0, 16);
}

export function makeRouteStopKey(poiId: string): string {
  const pid = poiId.trim();
  if (!pid) throw new Error("poi_id required");
  return `poi:${pid}`;
}

export function parseRouteStopKey(itemKey: string): string | null {
  const key = itemKey.trim();
  if (key.startsWith("poi:") && key.length > 4) {
    return key.slice(4);
  }
  return null;
}
